src/games/strategy/triplea/ai/proAI/util/ProOddsCalculator.java

```diff
@@ -0,0 +1,173 @@
+
+    final ProBattleResult result = checkIfNoAttackersOrDefenders(t, attackingUnits, defendingUnits);
+    if (result != null) {
+      return result;
+    }
+
+    // Determine if defenders have no chance
+    final double strengthDifference = ProBattleUtils.estimateStrengthDifference(t, attackingUnits, defendingUnits);
+    if (strengthDifference > 55) {
+      final boolean isLandAndCanOnlyBeAttackedByAir = !t.isWater() && Match.allMatch(attackingUnits, Matches.UnitIsAir);
+      return new ProBattleResult(100 + strengthDifference, 999 + strengthDifference, !isLandAndCanOnlyBeAttackedByAir,
+          attackingUnits, new ArrayList<Unit>(), 1);
+    }
+    return callBattleCalculator(player, t, attackingUnits, defendingUnits, bombardingUnits);
+  }
+
+  public ProBattleResult calculateBattleResults(final PlayerID player, final Territory t,
+      final List<Unit> attackingUnits, final List<Unit> defendingUnits, final Set<Unit> bombardingUnits,
+      final boolean isAttacker) {
+
+    final ProBattleResult result = checkIfNoAttackersOrDefenders(t, attackingUnits, defendingUnits);
+    if (result != null) {
+      return result;
+    }
+    return callBattleCalculator(player, t, attackingUnits, defendingUnits, bombardingUnits);
+  }
+
+  private ProBattleResult checkIfNoAttackersOrDefenders(final Territory t, final List<Unit> attackingUnits,
+      final List<Unit> defendingUnits) {
+    final GameData data = ProData.getData();
+
+    final boolean hasNoDefenders = Match.noneMatch(defendingUnits, Matches.UnitIsNotInfrastructure);
+    final boolean isLandAndCanOnlyBeAttackedByAir = !t.isWater() && Match.allMatch(attackingUnits, Matches.UnitIsAir);
+    if (attackingUnits.size() == 0) {
+      return new ProBattleResult();
+    } else if (hasNoDefenders && isLandAndCanOnlyBeAttackedByAir) {
+      return new ProBattleResult();
+    } else if (hasNoDefenders) {
+      return new ProBattleResult(100, 0.1, true, attackingUnits, new ArrayList<Unit>(), 0);
+    } else if (Properties.getSubRetreatBeforeBattle(data) && Match.allMatch(defendingUnits, Matches.UnitIsSub)
+        && Match.noneMatch(attackingUnits, Matches.UnitIsDestroyer)) {
+      return new ProBattleResult();
+    }
+    return null;
+  }
+
+
+  public ProBattleResult callBattleCalculator(final PlayerID player, final Territory t,
+      final List<Unit> attackingUnits, final List<Unit> defendingUnits, final Set<Unit> bombardingUnits) {
+    return callBattleCalculator(player, t, attackingUnits, defendingUnits, bombardingUnits, false);
+  }
+
+  public ProBattleResult callBattleCalculator(final PlayerID player, final Territory t,
+      final List<Unit> attackingUnits, final List<Unit> defendingUnits, final Set<Unit> bombardingUnits,
+      final boolean retreatWhenOnlyAirLeft) {
+    final GameData data = ProData.getData();
+
+    if (isCanceled || attackingUnits.isEmpty() || defendingUnits.isEmpty()) {
```

Comment:
I believe its due to different threads. The stop game is called by the UI thread when the user exits the game. The AI is running in a separate thread (otherwise you wouldn't be able to move around while the AI is thinking). The point is to abort any future calls to this method after stop game is called because this is a heavy battle calc method.